### PR TITLE
Add missing mini-expansion key item trading logic to Squintrox Dryeyes

### PR DESCRIPTION
**Pull request by MowFord**
**_I affirm:_**
- [x] I understand that if I do not agree to the following points by completing the checkboxes my PR will be ignored.
- [x] I understand I should leave resolving conversations to the LandSandBoat team so that reviewers won't miss what was said.
- [x] I have read and understood the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/LandSandBoat/server/blob/base/CODE_OF_CONDUCT.md).
- [x] I have _**tested my code and the things my code has changed**_ since the last commit in the PR and will test after any later commits.

## What does this pull request do?

This is the partner to my Tenshodo Coffer PR. Squintrox logic is not fun to look at in this first commit, but I hope to utilize the setbit logic from tenshodo coffer and build a couple tables to shorten the core logic. 

It is tested and technically works, but I am not satisfied with the layout, just putting it here to get feedback on a path to the end result.

Edit: Testing is complete and I'm satisfied with the behavior

Notes about the logic for this NPC:
 - The order of page 1 `onTrigger` does NOT match the order of event update options and event finish options, so this is manually defined per option value. 
 - The order of page 2 `onTrigger` DOES match the order of event finish options, so this was simple to map `option - 299` to the `menuMetadata` index

## Steps to test these changes

This one is tricky to test as there's multiple layers of logic here:
- Trades:
  - Trading the required items for: `CRIMSON_KEY`, `WHITE_CORAL_KEY`, and `MOOGLE_KEY` are the simplest. 
  - Trade should only succeed if the `charVar` for that particular key item maps to a timestamp on the player's charvar less than now (this value is set to next jst midnight upon a successful trade)
- Page 1 key items:
  - Page 1 lists the key items available based on
    - The respective mini expansion being complete (mission log shows the `FIN` mission as active)
    - the corresponding timestamp in the player's `charVar` player variable.
    - player not currently having the key item
  - if the player chooses that key item, the requirements are checked:
    - if the player has all required key items (most require full list, `BIRD_KEY` only requires 3 of the 6 key items) they receive the key item and all req (and prereq) key items are removed
    - if the player doesn't meet the requirements, info is given:
      - for all but `BLACK_CORAL_KEY` and `MOOGLE_KEY`, a simple text is given stating the requirements for obtaining the repeat key item
        - `MOOGLE_KEY`: the particular enfeeblement kit required is chosen/listed
        - `BLACK_CORAL_KEY`: the particular zone for digging is chosen/listed
- Page 2 key items:
  - These are all prerequisite key items for completing the mini-expansion battles again to obtain the key item rewards
  - They are not displayed if any is true
    - The respective mini expansion is not complete (mission log shows the `FIN` mission as active)
    - you currently have _ALL_ the key items 
    - you're on cooldown due to the `charVar` timestamp on your player
  - If chosen, they cost a varied amount of gil/beastmen seals, with custom messages for each type of result

Various screenshots of testing

image

image

image

### CODE REVIEW

File: scripts/zones/Port_Jeuno/npcs/Squintrox_Dryeyes.lua
Comment (zach2good) [inline scripts/zones/Port_Jeuno/npcs/Squintrox_Dryeyes.lua]: This looks suspiciously like it could be broken out into a table and then iterated over

```lua
-- { expansion, keyItem, maskValue }
local lookup =
{
    { <mark ACP condition somehow>, xi.ki.CRIMSON_KEY, bit.lshift(1, 1) },
    ...
}

local arg1 = 0
for _, entry in pairs(lookup) do
    local hasCompletedExpansion = player:checkCompleteEtc...(entry[1])
    local hasRelevantKeyItem        = player:hasKI...(entry[2])
    if hasCompletedExpansion or hasRelevantKeyItem then
        local maskValue = entry[3]
        arg1 = arg1 + maskValue
    end
end
```

File: scripts/zones/Port_Jeuno/npcs/Squintrox_Dryeyes.lua
Comment (zach2good) [inline scripts/zones/Port_Jeuno/npcs/Squintrox_Dryeyes.lua]: These two aren't used in here

File: scripts/zones/Port_Jeuno/npcs/Squintrox_Dryeyes.lua
Comment (zach2good) [inline scripts/zones/Port_Jeuno/npcs/Squintrox_Dryeyes.lua]: With the different key trades, you can probably write some helper funcs to deal with the repeated logic

File: scripts/zones/Port_Jeuno/npcs/Squintrox_Dryeyes.lua
Comment (zach2good) [inline scripts/zones/Port_Jeuno/npcs/Squintrox_Dryeyes.lua]: This giant block of things defined at the top is quite unsightly, but this is quite an overloaded NPC - so the option are to use these blocks inline if they're only used one (like here), or to break them out into local functions that takes the trade and/or the player as args and returns a bool

File: scripts/zones/Port_Jeuno/npcs/Squintrox_Dryeyes.lua
Comment (zach2good) [inline scripts/zones/Port_Jeuno/npcs/Squintrox_Dryeyes.lua]: I don't know the current state of ASA in LSB, but this will eventually be replaced by the numbered vars of the interaction framework, so I'd look into what those will look like for their eventual migration

File: scripts/zones/Port_Jeuno/npcs/Squintrox_Dryeyes.lua
Comment (zach2good) [inline scripts/zones/Port_Jeuno/npcs/Squintrox_Dryeyes.lua]: This extraction doesn't save much, just keep the usage inline

File: scripts/zones/Port_Jeuno/npcs/Squintrox_Dryeyes.lua
Comment (zach2good) [inline scripts/zones/Port_Jeuno/npcs/Squintrox_Dryeyes.lua:405]: Each of the blocks in here are quite unique, and take up quite varying amounts of space - I'd probably break all of them out into their own functions

File: scripts/zones/Port_Jeuno/npcs/Squintrox_Dryeyes.lua
Comment (zach2good) [inline scripts/zones/Port_Jeuno/npcs/Squintrox_Dryeyes.lua]: I'd look into the pattern I mentioned before about this being a data table that's then iterated over. Remember you can nest tables, so you can have a table of KI requirements as an entry in the main table

File: scripts/zones/Port_Jeuno/npcs/Squintrox_Dryeyes.lua
Comment (zach2good) [inline scripts/zones/Port_Jeuno/npcs/Squintrox_Dryeyes.lua]: For things like this, I'd define the table first and then use it, rather than inline in the loop definition

File: scripts/zones/Port_Jeuno/npcs/Squintrox_Dryeyes.lua
Comment (zach2good) [inline scripts/zones/Port_Jeuno/npcs/Squintrox_Dryeyes.lua]: For things like this, I'd define the table first and then use it, rather than inline in the loop definition.

Think about how these sorts of things are specific to the expansion they come from. Can they be added to the helpers files for those expansions and then referenced here like `xi.asaHelpers.keyItemsWhatever`?

File: scripts/zones/Port_Jeuno/npcs/Squintrox_Dryeyes.lua
Comment (MowFord) [reply scripts/zones/Port_Jeuno/npcs/Squintrox_Dryeyes.lua]: Fair point

File: scripts/zones/Port_Jeuno/npcs/Squintrox_Dryeyes.lua
Comment (MowFord) [reply scripts/zones/Port_Jeuno/npcs/Squintrox_Dryeyes.lua]: Good call I copy pasted definitions for consistent naming

File: scripts/zones/Port_Jeuno/npcs/Squintrox_Dryeyes.lua
Comment (MowFord) [reply scripts/zones/Port_Jeuno/npcs/Squintrox_Dryeyes.lua]: The real lpt is always in the comments

File: scripts/zones/Port_Jeuno/npcs/Squintrox_Dryeyes.lua
Comment (MowFord) [reply scripts/zones/Port_Jeuno/npcs/Squintrox_Dryeyes.lua]: Figured it was better to define all the complex requirements in one place at the top, but I could put them directly in the if blocks if that's preferred

File: scripts/zones/Port_Jeuno/npcs/Squintrox_Dryeyes.lua
Comment (MowFord) [reply scripts/zones/Port_Jeuno/npcs/Squintrox_Dryeyes.lua]: I figured this would just be a placeholder until that happens (it'll likely be me and flibe doing them)

File: scripts/zones/Port_Jeuno/npcs/Squintrox_Dryeyes.lua
Comment (MowFord) [reply scripts/zones/Port_Jeuno/npcs/Squintrox_Dryeyes.lua]: I think that is the route I will take. The same data is also used in the next event function in the same order so it should extend nicely

File: scripts/zones/Port_Jeuno/npcs/Squintrox_Dryeyes.lua
Comment (MowFord) [reply scripts/zones/Port_Jeuno/npcs/Squintrox_Dryeyes.lua]: I like it

File: scripts/zones/Port_Jeuno/npcs/Squintrox_Dryeyes.lua
Comment (MowFord) [reply scripts/zones/Port_Jeuno/npcs/Squintrox_Dryeyes.lua]: Oh interesting idea. I think I'd rather keep this file self contained tbh though. Since except for a could char var checks it's mostly self contained

Would it be ok to define the single table in this file with all of it and then update later when we audit the expansions?

In doing the expansions on wings it was very frustrating not having this file fleshed out first

**Comment by MowFord** [comment]
I have learned to run CI checks locally (go me), I also plan on investigating the commit ammending/force push to not require squashing every PR of mine. I don't mind if you squash this one or I can go back and combine the two commits myself.

I hope i didn't get too in my head with the `menuMetadata` table of key item info. I think I struck a balance between not repeating code and readability.

Note that I _have not_ checked the "i'm fully tested this code" box as I'm waiting on feedback on the formatting before I thoroughly check every iteration.

File: scripts/zones/Port_Jeuno/npcs/Squintrox_Dryeyes.lua
Comment (zach2good) [inline scripts/zones/Port_Jeuno/npcs/Squintrox_Dryeyes.lua]: The arbitrary gaps between the statement and the and/or keywords aren't needed here. Only put in extraneous whitespace if you're aligning things to something: like variable names or the = operators

File: scripts/zones/Port_Jeuno/npcs/Squintrox_Dryeyes.lua
Comment (zach2good) [inline scripts/zones/Port_Jeuno/npcs/Squintrox_Dryeyes.lua]: If you're not using i, you should use _ to signal that the key isn't being used

File: scripts/zones/Port_Jeuno/npcs/Squintrox_Dryeyes.lua
Comment (zach2good) [inline scripts/zones/Port_Jeuno/npcs/Squintrox_Dryeyes.lua]: This table is nice, but the expansion, charVar and reqItemCharVar keys aren't correctly aligned like the rest of the table - choose one style and stick to it

**Review by zach2good — COMMENTED** [review]
Liking the new layout 👍

File: scripts/zones/Port_Jeuno/npcs/Squintrox_Dryeyes.lua
Comment (MowFord) [reply scripts/zones/Port_Jeuno/npcs/Squintrox_Dryeyes.lua]: Ok so just

```
Statement and
Longer statement and
Short and
Blah
```
?

Comment (zach2good) [reply scripts/zones/Port_Jeuno/npcs/Squintrox_Dryeyes.lua]: Yessir

File: scripts/zones/Port_Jeuno/npcs/Squintrox_Dryeyes.lua
Comment (zach2good) [inline scripts/zones/Port_Jeuno/npcs/Squintrox_Dryeyes.lua]: i -> _

Comment (MowFord) [reply scripts/zones/Port_Jeuno/npcs/Squintrox_Dryeyes.lua]: :( sorry

**Review by zach2good — APPROVED** [review]

File: scripts/zones/Port_Jeuno/npcs/Squintrox_Dryeyes.lua
Comment (claywar) [inline scripts/zones/Port_Jeuno/npcs/Squintrox_Dryeyes.lua]: Debug print can be removed

File: scripts/zones/Port_Jeuno/npcs/Squintrox_Dryeyes.lua
Comment (claywar) [inline scripts/zones/Port_Jeuno/npcs/Squintrox_Dryeyes.lua]: These two lines can be replaced by the npcUtil function (which can handle either single input or tables of KIs)

File: scripts/zones/Port_Jeuno/npcs/Squintrox_Dryeyes.lua
Comment (claywar) [inline scripts/zones/Port_Jeuno/npcs/Squintrox_Dryeyes.lua:510]: Not as familiar with this NPC, but does he really pull from stored seals and don't require a trade?

Comment (MowFord) [reply scripts/zones/Port_Jeuno/npcs/Squintrox_Dryeyes.lua:510]: Sure does!

https://ffxiclopedia.fandom.com/wiki/Talk:Squintrox_Dryeyes

> She doesn't take the beastmen seals from you, she takes them from [Shami](https://ffxiclopedia.fandom.com/wiki/Shami). That's why she mentions that she and Shami have a deal. Trade the seals to him and you'll be able to buy the key items from Squintrox. --[Kyrie](https://ffxiclopedia.fandom.com/wiki/User:Kyrial) 02:28, June 30, 2010 (UTC)

File: scripts/zones/Port_Jeuno/npcs/Squintrox_Dryeyes.lua
Comment (MowFord) [reply scripts/zones/Port_Jeuno/npcs/Squintrox_Dryeyes.lua]: Good catch.

**Comment by MowFord** [comment]
I've updated the things clay found, but decided not to force push since zach already approved it

**Comment by zach2good** [comment]
Don't worry about force pushing things etc. it doesn't dismiss the approval (I think) and GitHub is pretty good at showing `New changes since you last viewed` and even keeps copies of earlier commits.

We're also pretty good at not yeeting merges in if other people have ongoing conversations

File: scripts/zones/Port_Jeuno/npcs/Squintrox_Dryeyes.lua
Comment (zach2good) [reply scripts/zones/Port_Jeuno/npcs/Squintrox_Dryeyes.lua:510]: And still does:
image

**Comment by MowFord** [comment]
Sorry to hijack a different PR but I just wanted to pose a quick question about the AMK work we'll be starting up.

Once this PR is merged I'd like to update the feature branch with `base` (to capture this and tenshodo coffer updates). Would you like me to `git merge base` for my first PR or would you like to update the feature branch before I submit the first PR? I'm not well versed in the best ways to handle merge commits.

**Comment by zach2good** [comment]
Once this merges I'll rebase the feature branch, so you can start targetting that without having to do any merges 👍

**Pull request merged**
